### PR TITLE
Feature/773 negation rule on bool

### PR DESCRIPTION
**Pull request by Zejnilovic**
Allow negation rule to be used on booleans

With great help from @yruslan

### CODE REVIEW

File: utils/src/main/scala/za/co/absa/enceladus/utils/validation/SchemaPathValidator.scala [hunk 1]
```diff
@@ -108,6 +108,20 @@ object SchemaPathValidator {
...
+    * Validate schema path data type is algebraic
+    *
+    * @param schema      A Spark schema
+    * @param fieldPath   A path to a field (e.g. "data.employees.employee.age")
+    * @return A list of ValidationErrors objects, each containing a column name and the list of errors and warnings
+    */
+  def validateSchemaPathAlgebraic(schema: StructType, fieldPath: String): Seq[ValidationIssue] = {
+    validateSchemaPathType(schema, fieldPath) {
+      case _: NumericType |  _: BooleanType => Seq.empty[ValidationIssue]
+      case k => Seq(ValidationError(s"The datatype '${k.typeName}' of '$fieldPath' field is not a algebraic type"))
```
Comment (yruslan) [inline utils/src/main/scala/za/co/absa/enceladus/utils/validation/SchemaPathValidator.scala]: The error message might be confusing. `Algebraic type` is a not very common notion. I would rephrase it to be something like `The datatype '${k.typeName}' of '$fieldPath' field is neither numeric nor boolean`

File: conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala [hunk 2]
```diff
@@ -39,14 +39,19 @@ class NegationRuleSuite extends FunSuite with SparkTestBase with LoggerTestBase{
...
-    }.getMessage contains "field is not a numeric type")
+    }.getMessage contains "field is not a algebraic type")
   }
 
   test("Negation conformance rule should negate positive numeric values") {
     val inputDataset = NegationRuleSamples.Positive.data.toDS
     val expectedDataset = NegationRuleSamples.dataset
     val expectedJSON = NegationRuleSamples.Positive.conformedJSON
 
+//    inputDataset.show(false)
```
Comment (yruslan) [inline conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala]: Redundant commnt

File: conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala [hunk 3]
```diff
@@ -39,14 +39,19 @@ class NegationRuleSuite extends FunSuite with SparkTestBase with LoggerTestBase{
 
     assert(intercept[ValidationException] {
       NegationRuleInterpreter.validateInputField("dataset", schema, "date")
-    }.getMessage contains "field is not a numeric type")
+    }.getMessage contains "field is not a algebraic type")
```
Comment (yruslan) [inline conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala]: Just a reminder. If you going to change the error message of the exception do not forget to change it here as well.

File: conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala (hunk 2 above)
Comment (Zejnilovic) [reply conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala]: This is in general just garbage left there by me.

File: conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala (hunk 3 above)
Comment (Zejnilovic) [reply conformance/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleSuite.scala]: I can promise you, I would forget. Thanks

File: utils/src/main/scala/za/co/absa/enceladus/utils/validation/SchemaPathValidator.scala [hunk 4]
```diff
@@ -108,6 +108,21 @@ object SchemaPathValidator {
...
+    *
+    * @param schema      A Spark schema
+    * @param fieldPath   A path to a field (e.g. "data.employees.employee.age")
+    * @return A list of ValidationErrors objects, each containing a column name and the list of errors and warnings
+    */
+  def validateSchemaPathAlgebraic(schema: StructType, fieldPath: String): Seq[ValidationIssue] = {
+    validateSchemaPathType(schema, fieldPath) {
+      case _: NumericType |  _: BooleanType => Seq.empty[ValidationIssue]
+      case k => Seq(ValidationError(s"The datatype '${k.typeName}' of '$fieldPath' field " +
+        "is neither NumericType or BooleanType"))
```
Comment (benedeki) [inline utils/src/main/scala/za/co/absa/enceladus/utils/validation/SchemaPathValidator.scala]:
```suggestion
        "is neither NumericType nor BooleanType"))
```

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala [hunk 5]
```diff
@@ -40,8 +40,10 @@ case class NegationRuleInterpreter(rule: NegationConformanceRule) extends RuleIn
       case _: DecimalType =>
         // Negating decimal cannot fail
         DeepArrayTransformations.nestedWithColumnMap(df, rule.inputColumn, rule.outputColumn, c => negate(c))
-      case _: DoubleType | _: FloatType =>
-        // Negating floating point numbers cannot fail, but we need to account for signed zeros (see the note for getNegator()).
+      case _: DoubleType | _: FloatType | _: BooleanType =>
+        // Negating boolean values and floating point numbers cannot fail. But:
```
Comment (benedeki) [inline conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: The evaluation is already somewhat chaotic, and repeatedly test for `fieldType`. Not requesting a full refactor, but could you perhaps make the `BooleanType` a separate `case` similarly to `DecimalType` with the `not(c)` used directly?

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala [hunk 6]
```diff
@@ -78,7 +81,8 @@ case class NegationRuleInterpreter(rule: NegationConformanceRule) extends RuleIn
       case _: ShortType => when(inputColumn === Short.MinValue, errorColumnUDF).otherwise(null)
       case _: IntegerType => when(inputColumn === Int.MinValue, errorColumnUDF).otherwise(null)
       case _: LongType => when(inputColumn === Long.MinValue, errorColumnUDF).otherwise(null)
-      case _ => null
+      case a => throw new IllegalArgumentException("NegationRuleInterpreter.getError() should be called only for " +
```
Comment (benedeki) [inline conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Why this change? Due to the nested checking this branch is unreachable anyway. And if reached, will halt the whole conformance at runtime in the middle of processing (which is somewhat late anyway). 🤔

Comment (Zejnilovic) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Bad error message from this part of code is, what I think stopped me and I needed to ask Ruslan for help. I vote we keep the message. Otherwise, you get unreadable messages.

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala [hunk 7]
```diff
@@ -40,8 +40,10 @@ case class NegationRuleInterpreter(rule: NegationConformanceRule) extends RuleIn
       case _: DecimalType =>
         // Negating decimal cannot fail
         DeepArrayTransformations.nestedWithColumnMap(df, rule.inputColumn, rule.outputColumn, c => negate(c))
-      case _: DoubleType | _: FloatType =>
-        // Negating floating point numbers cannot fail, but we need to account for signed zeros (see the note for getNegator()).
+      case _: DoubleType | _: FloatType | _: BooleanType =>
+        // Negating boolean values and floating point numbers cannot fail. But:
+        // * for floating point data types we need to account for signed zeros (see the note for getNegator()).
+        // * for booleant data type we need to use different negation function.
```
[the remark this answers is not in the document]
Comment (Zejnilovic) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Overwritten as a whole. Text no longer exists

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala (hunk 5 above)
Comment (Zejnilovic) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Refactored a little. Please check

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala (hunk 6 above)
Comment (yruslan) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Exactly. That branch should be unreachable. But when a new data type is added and the function returns null the error message we get doesn't make any sense. It is a NullPointerException somewhere deep in DeepArrayTransformation. Double deep 😄

Comment (benedeki) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Shouldn't there be an error message there then instead of crashing the whole process?

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala [hunk 8]
```diff
@@ -29,6 +29,7 @@ import za.co.absa.enceladus.utils.validation.SchemaPathValidator
 
 case class NegationRuleInterpreter(rule: NegationConformanceRule) extends RuleInterpreter {
 
+  //noinspection ScalaStyle
```
Comment (benedeki) [inline conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Why this broad switch off of the Scalastyle check? 😟

Comment (Zejnilovic) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: I think I did something to my IDE 😑

File: conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala (hunk 6 above)
Comment (benedeki) [reply conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/NegationRuleInterpreter.scala]: Agreed killing the processing is good in this case of _"should never happen"_ 😉
